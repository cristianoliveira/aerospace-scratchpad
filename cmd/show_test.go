package cmd

import (
	"fmt"
	"strings"
	"testing"

	"github.com/cristianoliveira/aerospace-marks/pkgs/aerospacecli"
	"github.com/cristianoliveira/aerospace-scratchpad/internal/mocks/aerospacecli"
	"github.com/cristianoliveira/aerospace-scratchpad/internal/testutils"
	"github.com/gkampitakis/go-snaps/snaps"
	"go.uber.org/mock/gomock"
)

func TestFunction(t *testing.T) {
	t.Run("moves the current focused window to scratchpad when empty", func(t *testing.T) {
		command := "show"
		args := []string{command, ""}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
					{
						AppName:  "Finder",
						WindowID: 5678,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},

				FocusedWindowId: 5678,
			},
		}

		allWindows := testutils.ExtractAllWindows(tree)
		focusedTree := testutils.ExtractFocusedTree(tree)
		focusedWindow := testutils.ExtractFocusedWindow(tree)

		aerospaceClient := aerospacecli_mock.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetFocusedWindow().
				Return(focusedWindow, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(allWindows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(focusedTree.Workspace, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindowsByWorkspace(focusedTree.Workspace.Workspace).
				Return(focusedTree.Windows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetFocusedWindow().
				Return(focusedWindow, nil).
				Times(1),

			aerospaceClient.EXPECT().
				MoveWindowToWorkspace(focusedWindow.WindowID, "scratchpad").
				Return(nil).
				Times(1),

			aerospaceClient.EXPECT().
				SetLayout(focusedWindow.WindowID, "floating").
				Return(nil).
				Times(1),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err != nil {
			t.Errorf("Expected no error, got %v", err)
		}

		if out == "" {
			t.Errorf("Expected output, got empty string")
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		expectedError := fmt.Sprintf("Error\n%+v", err)
		snaps.MatchSnapshot(t, tree, cmdAsString, "Output", out, expectedError)
	})

	t.Run(
		"set focus to window if already in the focused workspace but not focused",
		func(t *testing.T) {
			command := "show"
			args := []string{command, "Finder"}

			ctrl := gomock.NewController(t)
			defer ctrl.Finish()

			tree := []testutils.AeroSpaceTree{
				{
					Windows: []aerospacecli.Window{
						{
							AppName:  "Notepad",
							WindowID: 1234,
						},
						{
							AppName:  "Finder",
							WindowID: 5678,
						},
					},
					Workspace: &aerospacecli.Workspace{
						Workspace: "ws1",
					},

					FocusedWindowId: 1234,
				},
			}

			allWindows := testutils.ExtractAllWindows(tree)
			focusedTree := testutils.ExtractFocusedTree(tree)
			focusedWindow := testutils.ExtractFocusedWindow(tree)

			aerospaceClient := aerospacecli_mock.NewMockAeroSpaceClient(ctrl)
			gomock.InOrder(
				aerospaceClient.EXPECT().
					GetAllWindows().
					Return(allWindows, nil).
					Times(1),

				aerospaceClient.EXPECT().
					GetFocusedWorkspace().
					Return(focusedTree.Workspace, nil).
					Times(1),

				aerospaceClient.EXPECT().
					GetAllWindowsByWorkspace(focusedTree.Workspace.Workspace).
					Return(focusedTree.Windows, nil).
					Times(1),

				aerospaceClient.EXPECT().
					GetFocusedWindow().
					Return(focusedWindow, nil).
					Times(1),

				aerospaceClient.EXPECT().
					SetFocusByWindowID(focusedTree.Windows[1].WindowID).
					Return(nil).
					Times(1),

				// DO NOT set the layout to floating
				aerospaceClient.EXPECT().
					SetLayout(gomock.Any(), "floating").
					Return(nil).
					Times(0),
			)

			cmd := RootCmd(aerospaceClient)
			out, err := testutils.CmdExecute(cmd, args...)
			if err != nil {
				t.Errorf("Expected no error, got %v", err)
			}

			if out == "" {
				t.Errorf("Expected output, got empty string")
			}

			cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
			expectedError := fmt.Sprintf("Error\n%+v", err)
			snaps.MatchSnapshot(t, tree, cmdAsString, "Output", out, expectedError)
		})

	t.Run("moves a window to scratchpad by pattern", func(t *testing.T) {
		command := "show"
		args := []string{command, "Finder"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
					{
						AppName:  "Finder",
						WindowID: 5678,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},

				FocusedWindowId: 5678,
			},
		}

		allWindows := testutils.ExtractAllWindows(tree)
		focusedTree := testutils.ExtractFocusedTree(tree)
		focusedWindow := testutils.ExtractFocusedWindow(tree)

		aerospaceClient := aerospacecli_mock.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(allWindows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(focusedTree.Workspace, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindowsByWorkspace("ws1").
				Return(focusedTree.Windows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetFocusedWindow().
				Return(focusedWindow, nil).
				Times(1),

			aerospaceClient.EXPECT().
				MoveWindowToWorkspace(focusedWindow.WindowID, "scratchpad").
				Return(nil).
				Times(1),

			aerospaceClient.EXPECT().
				SetFocusByWindowID(focusedWindow.WindowID).
				Return(nil).
				Times(0),

			// When moving to scratchpad, set the layout to floating
			aerospaceClient.EXPECT().
				SetLayout(focusedWindow.WindowID, "floating").
				Return(nil).
				Times(1),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err != nil {
			t.Errorf("Expected no error, got %v", err)
		}

		if out == "" {
			t.Errorf("Expected output, got empty string")
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		expectedError := fmt.Sprintf("Error\n%+v", err)
		snaps.MatchSnapshot(t, tree, cmdAsString, "Output", out, expectedError)
	})

	t.Run("summon the window to the current workspace if in another workspace", func(t *testing.T) {
		command := "show"
		args := []string{command, "Finder"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
					{
						AppName:  "Finder",
						WindowID: 5678,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},
				FocusedWindowId: 0, // Not focused
			},
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Terminal",
						WindowID: 91011,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws2",
				},
				FocusedWindowId: 91011,
			},
		}

		allWindows := testutils.ExtractAllWindows(tree)
		focusedTree := testutils.ExtractFocusedTree(tree)
		focusedWindow := testutils.ExtractFocusedWindow(tree)

		aerospaceClient := aerospacecli_mock.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(allWindows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(focusedTree.Workspace, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindowsByWorkspace(focusedTree.Workspace.Workspace).
				Return(focusedTree.Windows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				MoveWindowToWorkspace(
					tree[0].Windows[1].WindowID,
					focusedTree.Workspace.Workspace).
				Return(nil).
				Times(1),

			aerospaceClient.EXPECT().
				SetFocusByWindowID(
					tree[0].Windows[1].WindowID).
				Return(nil).
				Times(1),

			// When moving to scratchpad, set the layout to floating
			aerospaceClient.EXPECT().
				SetLayout(focusedWindow.WindowID, "floating").
				Return(nil).
				Times(0),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err != nil {
			t.Errorf("Expected no error, got %v", err)
		}

		if out == "" {
			t.Errorf("Expected output, got empty string")
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		expectedError := fmt.Sprintf("Error\n%+v", err)
		snaps.MatchSnapshot(t, tree, cmdAsString, "Output", out, expectedError)
	})
}
