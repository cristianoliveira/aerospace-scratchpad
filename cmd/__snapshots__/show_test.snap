
[TestShowCmd/fails_when_pattern_is_empty - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:1234, WindowTitle:"", AppName:"Notepad", AppBundleID:"", Workspace:""},
            {WindowID:5678, WindowTitle:"", AppName:"Finder", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowID: 5678,
    },
}
aerospace-scratchpad show 

Output

Error
Error: <pattern> cannot be empty

---

[TestShowCmd/fails_when_pattern_doesn_match_any_window - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:1234, WindowTitle:"", AppName:"Notepad", AppBundleID:"", Workspace:""},
            {WindowID:5678, WindowTitle:"", AppName:"Finder", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowID: 1234,
    },
}
aerospace-scratchpad show foo

Output

Error
Error: no windows matched the pattern 'foo'

---

[TestShowCmd/set_focus_to_window_if_already_in_the_focused_workspace_but_not_focused - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:1234, WindowTitle:"", AppName:"Notepad", AppBundleID:"", Workspace:""},
            {WindowID:5678, WindowTitle:"", AppName:"Finder", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowID: 1234,
    },
}
aerospace-scratchpad show Finder

Output
Window '5678 | Finder ' is focused

Error
<nil>
---

[TestShowCmd/moves_a_window_to_scratchpad_by_pattern - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:1234, WindowTitle:"", AppName:"Notepad", AppBundleID:"", Workspace:""},
            {WindowID:5678, WindowTitle:"", AppName:"Finder", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowID: 5678,
    },
}
aerospace-scratchpad show Finder

Output
Window '5678 | Finder ' hidden to scratchpad

Error
<nil>
---

[TestShowCmd/summon_the_window_to_the_current_workspace_if_in_another_workspace - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:1234, WindowTitle:"", AppName:"Notepad", AppBundleID:"", Workspace:""},
            {WindowID:5678, WindowTitle:"", AppName:"Finder", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowID: 0,
    },
    {
        Windows: {
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowID: 91011,
    },
}
aerospace-scratchpad show Finder

Output
Window '5678 | Finder ' is moved to workspace 'ws2'

Error
<nil>
---

[TestShowCmd/MultipleWindows/brings_all_windows_to_focused_workspace - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:5678, WindowTitle:"", AppName:"Finder1", AppBundleID:"", Workspace:"ws1"},
            {WindowID:5679, WindowTitle:"", AppName:"Finder2", AppBundleID:"", Workspace:"ws1"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowID: 0,
    },
    {
        Windows: {
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:"ws2"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowID: 91011,
    },
}
aerospace-scratchpad show Finder

Output
Window '5678 | Finder1  | ws1' is moved to workspace 'ws2'
Window '5679 | Finder2  | ws1' is moved to workspace 'ws2'

Error
<nil>
---

[TestShowCmd/MultipleWindows/sends_all_windows_to_scratchpad_if_at_least_one_window_is_focused - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowID: 0,
    },
    {
        Windows: {
            {WindowID:5678, WindowTitle:"", AppName:"Finder1", AppBundleID:"", Workspace:"ws2"},
            {WindowID:5679, WindowTitle:"", AppName:"Finder2", AppBundleID:"", Workspace:"ws2"},
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:"ws2"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowID: 5678,
    },
}
aerospace-scratchpad show Finder

Output
Window '5678 | Finder1  | ws2' hidden to scratchpad
Window '5679 | Finder2  | ws2' hidden to scratchpad

Error
<nil>
---

[TestShowCmd/MultipleWindows/gives_priority_to_bringing_scratchpads_together - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:5678, WindowTitle:"", AppName:"Finder1", AppBundleID:"", Workspace:"ws1"},
            {WindowID:22, WindowTitle:"", AppName:"Browser", AppBundleID:"", Workspace:"ws1"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowID: 0,
    },
    {
        Windows: {
            {WindowID:5679, WindowTitle:"", AppName:"Finder2", AppBundleID:"", Workspace:"ws2"},
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:"ws2"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowID: 91011,
    },
}
aerospace-scratchpad show Finder

Output
Window '5678 | Finder1  | ws1' is moved to workspace 'ws2'
Window '5679 | Finder2  | ws2' is focused

Error
<nil>
---

[TestShowCmd/MultipleWindows/when_bringing_windows_together,_it_doesnt_change_focus - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:5678, WindowTitle:"", AppName:"Finder1", AppBundleID:"", Workspace:"ws1"},
            {WindowID:22, WindowTitle:"", AppName:"Browser", AppBundleID:"", Workspace:"ws1"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowID: 0,
    },
    {
        Windows: {
            {WindowID:5679, WindowTitle:"", AppName:"Finder2", AppBundleID:"", Workspace:"ws2"},
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:"ws2"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowID: 5679,
    },
}
aerospace-scratchpad show Finder

Output
Window '5678 | Finder1  | ws1' is moved to workspace 'ws2'
Window '5679 | Finder2  | ws2' is focused

Error
<nil>
---

[TestShowCmd/MultipleWindows/Filter_flag:_brings_any_windows_that_matches_filter - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:5678, WindowTitle:"Finder - foo and zas", AppName:"Finder1", AppBundleID:"", Workspace:"ws1"},
            {WindowID:5679, WindowTitle:"", AppName:"Finder2 - bar and baz", AppBundleID:"", Workspace:"ws1"},
        },
        Workspace:       &aerospace.Workspace{Workspace:".scratchpad"},
        FocusedWindowID: 0,
    },
    {
        Windows: {
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:"ws2"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowID: 91011,
    },
}
aerospace-scratchpad show Finder --filter window-title=foo

Output
Window '5678 | Finder1 | Finder - foo and zas | ws1' is moved to workspace 'ws2'

Error
<nil>
---

[TestShowCmd - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:5678, WindowTitle:"Finder - foo and zas", AppName:"Finder1", AppBundleID:"com.linux.finder", Workspace:"ws1"},
            {WindowID:5679, WindowTitle:"Finder2 - foo and baz", AppName:"Finder2", AppBundleID:"com.apple.finder", Workspace:"ws1"},
            {WindowID:5680, WindowTitle:"Finder2 - bar and baz", AppName:"Finder2", AppBundleID:"com.apple.finder", Workspace:"ws1"},
        },
        Workspace:       &aerospace.Workspace{Workspace:".scratchpad"},
        FocusedWindowId: 0,
    },
    {
        Windows: {
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:"ws2"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowId: 91011,
    },
}
aerospace-scratchpad show Finder -F window-title=foo -F app-bundle-id=linux

Output
Window '5678 | Finder1 | Finder - foo and zas | ws1 | com.linux.finder' is summoned

Error
<nil>
---

[TestShowCmd - 2]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:5678, WindowTitle:"Finder - foo and zas", AppName:"Finder1", AppBundleID:"com.linux.finder", Workspace:"ws1"},
            {WindowID:5679, WindowTitle:"Finder2 - foo and baz", AppName:"Finder2", AppBundleID:"com.apple.finder", Workspace:"ws1"},
            {WindowID:5680, WindowTitle:"Finder2 - bar and baz", AppName:"Finder2", AppBundleID:"com.apple.finder", Workspace:"ws1"},
        },
        Workspace:       &aerospace.Workspace{Workspace:".scratchpad"},
        FocusedWindowId: 0,
    },
    {
        Windows: {
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:"ws2"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowId: 91011,
    },
}
aerospace-scratchpad show Finder -F window-title=foo -F app-bundle-id=linux

Output
Window '5678 | Finder1 | Finder - foo and zas | ws1 | com.linux.finder' is summoned

Error
<nil>
---

[TestShowCmd/MultipleWindows/Filter_flag:_brings_any_windows_that_matches_filter_-_allow_multiple - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:5678, WindowTitle:"Finder - foo and zas", AppName:"Finder1", AppBundleID:"com.linux.finder", Workspace:"ws1"},
            {WindowID:5679, WindowTitle:"Finder2 - foo and baz", AppName:"Finder2", AppBundleID:"com.apple.finder", Workspace:"ws1"},
            {WindowID:5680, WindowTitle:"Finder2 - bar and baz", AppName:"Finder2", AppBundleID:"com.apple.finder", Workspace:"ws1"},
        },
        Workspace:       &aerospace.Workspace{Workspace:".scratchpad"},
        FocusedWindowID: 0,
    },
    {
        Windows: {
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:"ws2"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowID: 91011,
    },
}
aerospace-scratchpad show Finder -F window-title=foo -F app-bundle-id=linux

Output
Window '5678 | Finder1 | Finder - foo and zas | ws1 | com.linux.finder' is moved to workspace 'ws2'

Error
<nil>
---

[TestShowCmd/MultipleWindows/Filter_flag:_fails_when_unknown_filter_property_is_used - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:5678, WindowTitle:"", AppName:"Finder1", AppBundleID:"", Workspace:"ws1"},
            {WindowID:5670, WindowTitle:"", AppName:"Finder2", AppBundleID:"", Workspace:"ws1"},
        },
        Workspace:       &aerospace.Workspace{Workspace:".scratchpad"},
        FocusedWindowID: 5670,
    },
}
aerospace-scratchpad show Finder --filter unknown=foo

Output

Error
Error: error applying filters to window 'Finder1': unknown filter property: unknown

---

[TestShowCmd/MultipleWindows/Filter_flag:_shows_error_messages_when_no_match - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:5678, WindowTitle:"Finder - foo and zas", AppName:"Finder1", AppBundleID:"", Workspace:"ws1"},
            {WindowID:5679, WindowTitle:"", AppName:"Finder2 - bar and baz", AppBundleID:"", Workspace:"ws1"},
        },
        Workspace:       &aerospace.Workspace{Workspace:".scratchpad"},
        FocusedWindowID: 0,
    },
    {
        Windows: {
            {WindowID:91011, WindowTitle:"", AppName:"Terminal", AppBundleID:"", Workspace:"ws2"},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws2"},
        FocusedWindowID: 91011,
    },
}
aerospace-scratchpad show Finder --filter window-title=cantfindme

Output

Error
Error: no windows matched the pattern 'Finder' with the given filters

---
